File: src/commands/doctor.rs
use crate::calm_format;
use crate::config;
use anyhow::Result;
use colored::Colorize;
use std::env;
use std::path::Path;
use std::process::Command as Proc;

enum Status {
    Ok(String),
    Warn(String),
    Fail(String),
}

fn print_status(label: &str, status: Status) {
    let (icon, msg): (colored::ColoredString, String) = match status {
        Status::Ok(m) => ("✓".truecolor(166, 227, 161), m),
        Status::Warn(m) => ("!".truecolor(249, 226, 175), m),
        Status::Fail(m) => ("✗".truecolor(243, 139, 168), m),
    };
    println!("  {icon} {:<28} {}", label, msg.dimmed());
}

fn is_arch_based() -> Status {
    if let Ok(os_release) = std::fs::read_to_string("/etc/os-release") {
        let lower = os_release.to_lowercase();
        if lower.contains("endeavouros") {
            Status::Ok("EndeavourOS detected".into())
        } else if lower.contains("arch") {
            Status::Ok("Arch-based distro detected".into())
        } else {
            Status::Warn("not Arch-based — some features may not work".into())
        }
    } else {
        Status::Warn("/etc/os-release not found".into())
    }
}

fn command_exists(cmd: &str) -> bool {
    Proc::new("which")
        .arg(cmd)
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false)
}

fn check_git() -> Status {
    if command_exists("git") {
        Status::Ok("found".into())
    } else {
        Status::Fail("not found — install with `pacman -S git`".into())
    }
}

fn check_hyprland() -> Status {
    if env::var("HYPRLAND_INSTANCE_SIGNATURE").is_ok() || command_exists("hyprctl") {
        Status::Ok("Hyprland detected".into())
    } else {
        Status::Warn("not running under Hyprland".into())
    }
}

fn check_truecolor() -> Status {
    match env::var("COLORTERM") {
        Ok(v) if v.contains("truecolor") || v.contains("24bit") => {
            Status::Ok("truecolor supported".into())
        }
        _ => Status::Warn("COLORTERM not set — pastel palette may render inaccurately".into()),
    }
}

fn check_config_dir() -> Result<Status> {
    config::ensure_scaffold()?;
    let dir = config::config_dir()?;
    if Path::new(&dir).exists() {
        Ok(Status::Ok(dir.display().to_string()))
    } else {
        Ok(Status::Fail("could not create config dir".into()))
    }
}

fn check_calm_config() -> Status {
    match config::config_file().and_then(|p| calm_format::parse_file(&p)) {
        Ok(_) => Status::Ok("config.calm parses cleanly".into()),
        Err(e) => Status::Fail(format!("config.calm failed to parse: {e}")),
    }
}

fn check_editor() -> Status {
    match env::var("EDITOR") {
        Ok(e) => Status::Ok(e),
        Err(_) => Status::Warn("$EDITOR not set — `calm config` will fall back to nano".into()),
    }
}

pub fn run() -> Result<()> {
    println!("{}", "Calm-shell doctor".truecolor(203, 166, 247).bold());
    println!("{}", "Diagnosing system compatibility...\n".dimmed());

    print_status("OS", is_arch_based());
    print_status("git", check_git());
    print_status("Hyprland", check_hyprland());
    print_status("Terminal colors", check_truecolor());
    print_status("$EDITOR", check_editor());
    print_status("Config directory", check_config_dir()?);
    print_status("config.calm", check_calm_config());

    println!("\n{}", "Done.".dimmed());
    Ok(())
}

File: src/commands/doctor.c
#include "doctor.h"

#include <ctype.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/wait.h>
#include <unistd.h>

#include "../calm_format.h"
#include "../config.h"
#include "../util.h"

typedef enum { STATUS_OK, STATUS_WARN, STATUS_FAIL } StatusKind;

static void print_status(const char *label, StatusKind kind, const char *msg) {
    const char *icon;
    switch (kind) {
        case STATUS_OK:
            icon = "\x1b[38;2;166;227;161m\xE2\x9C\x93\x1b[0m";
            break;
        case STATUS_WARN:
            icon = "\x1b[38;2;249;226;175m!\x1b[0m";
            break;
        default:
            icon = "\x1b[38;2;243;139;168m\xE2\x9C\x97\x1b[0m";
            break;
    }
    printf("  %s %-28s \x1b[2m%s\x1b[0m\n", icon, label, msg);
}

static void check_is_arch_based(char *msg_out, size_t msg_cap, StatusKind *kind) {
    char *os_release = read_file_to_string("/etc/os-release");
    if (!os_release) {
        *kind = STATUS_WARN;
        snprintf(msg_out, msg_cap, "/etc/os-release not found");
        return;
    }
    for (char *p = os_release; *p; p++) {
        *p = (char)tolower((unsigned char)*p);
    }
    if (strstr(os_release, "endeavouros")) {
        *kind = STATUS_OK;
        snprintf(msg_out, msg_cap, "EndeavourOS detected");
    } else if (strstr(os_release, "arch")) {
        *kind = STATUS_OK;
        snprintf(msg_out, msg_cap, "Arch-based distro detected");
    } else {
        *kind = STATUS_WARN;
        snprintf(msg_out, msg_cap, "not Arch-based -- some features may not work");
    }
    free(os_release);
}

static bool command_exists(const char *cmd) {
    pid_t pid = fork();
    if (pid < 0) {
        return false;
    }
    if (pid == 0) {
        int devnull = open("/dev/null", O_WRONLY);
        if (devnull >= 0) {
            dup2(devnull, STDOUT_FILENO);
            dup2(devnull, STDERR_FILENO);
            close(devnull);
        }
        execlp("which", "which", cmd, (char *)NULL);
        _exit(127);
    }
    int status = 0;
    waitpid(pid, &status, 0);
    return WIFEXITED(status) && WEXITSTATUS(status) == 0;
}

static void check_git(char *msg_out, size_t msg_cap, StatusKind *kind) {
    if (command_exists("git")) {
        *kind = STATUS_OK;
        snprintf(msg_out, msg_cap, "found");
    } else {
        *kind = STATUS_FAIL;
        snprintf(msg_out, msg_cap, "not found -- install with `pacman -S git`");
    }
}

static void check_hyprland(char *msg_out, size_t msg_cap, StatusKind *kind) {
    const char *sig = getenv("HYPRLAND_INSTANCE_SIGNATURE");
    if ((sig && sig[0] != '\0') || command_exists("hyprctl")) {
        *kind = STATUS_OK;
        snprintf(msg_out, msg_cap, "Hyprland detected");
    } else {
        *kind = STATUS_WARN;
        snprintf(msg_out, msg_cap, "not running under Hyprland");
    }
}

static void check_truecolor(char *msg_out, size_t msg_cap, StatusKind *kind) {
    const char *v = getenv("COLORTERM");
    if (v && (strstr(v, "truecolor") || strstr(v, "24bit"))) {
        *kind = STATUS_OK;
        snprintf(msg_out, msg_cap, "truecolor supported");
    } else {
        *kind = STATUS_WARN;
        snprintf(msg_out, msg_cap, "COLORTERM not set -- pastel palette may render inaccurately");
    }
}

static bool check_config_dir(char *msg_out, size_t msg_cap, StatusKind *kind) {
    if (!ensure_scaffold()) {
        *kind = STATUS_FAIL;
        snprintf(msg_out, msg_cap, "could not create config dir");
        return true;
    }
    char *dir = config_dir();
    *kind = STATUS_OK;
    snprintf(msg_out, msg_cap, "%s", dir ? dir : "?");
    free(dir);
    return true;
}

static void check_calm_config(char *msg_out, size_t msg_cap, StatusKind *kind) {
    char *path = config_file();
    if (!path) {
        *kind = STATUS_FAIL;
        snprintf(msg_out, msg_cap, "could not resolve config path");
        return;
    }
    CalmDocument doc;
    char *err = NULL;
    if (calm_parse_file(path, &doc, &err)) {
        calm_document_free(&doc);
        *kind = STATUS_OK;
        snprintf(msg_out, msg_cap, "config.calm parses cleanly");
    } else {
        *kind = STATUS_FAIL;
        snprintf(msg_out, msg_cap, "config.calm failed to parse: %s", err ? err : "unknown error");
    }
    free(err);
    free(path);
}

static void check_editor(char *msg_out, size_t msg_cap, StatusKind *kind) {
    const char *e = getenv("EDITOR");
    if (e && e[0] != '\0') {
        *kind = STATUS_OK;
        snprintf(msg_out, msg_cap, "%s", e);
    } else {
        *kind = STATUS_WARN;
        snprintf(msg_out, msg_cap, "$EDITOR not set -- `calm config` will fall back to nano");
    }
}

int doctor_cmd_run(void) {
    printf("\x1b[38;2;203;166;247m\x1b[1mCalm-shell doctor\x1b[0m\n");
    printf("\x1b[2mDiagnosing system compatibility...\x1b[0m\n\n");

    char msg[512];
    StatusKind kind;

    check_is_arch_based(msg, sizeof(msg), &kind);
    print_status("OS", kind, msg);

    check_git(msg, sizeof(msg), &kind);
    print_status("git", kind, msg);

    check_hyprland(msg, sizeof(msg), &kind);
    print_status("Hyprland", kind, msg);

    check_truecolor(msg, sizeof(msg), &kind);
    print_status("Terminal colors", kind, msg);

    check_editor(msg, sizeof(msg), &kind);
    print_status("$EDITOR", kind, msg);

    check_config_dir(msg, sizeof(msg), &kind);
    print_status("Config directory", kind, msg);

    check_calm_config(msg, sizeof(msg), &kind);
    print_status("config.calm", kind, msg);

    printf("\n\x1b[2mDone.\x1b[0m\n");
    return 0;
}
